### ISSUE TITLE
V3-8 update-freq 4 and V3-32 training having very different behaviors than GPU

### ISSUE THREAD

**Issue by kkissmart** [pytorch/fairseq#2332]
I am running a transformer based mbart-base model.

I used args.tpu

I initially trained on V3-32, the loss doesnt go down. then I switch to V3-8 with update freq 4, and loss looks better but still very off from gpu logs.

TPU log:
2020-07-17 04:00:55 | INFO | train_inner | epoch 001:    100 / 2952000 loss=16.137, ppl=72076.5, wps=0, ups=0, wpb=1928, bsz=2, num_updates=100, lr=1e-06, gnorm=4.43, train_wall=13, wall=191
2020-07-17 04:00:55 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-17 04:01:26 | INFO | train_inner | epoch 001:    200 / 2952000 loss=14.498, ppl=23144.6, wps=60.1, ups=0.03, wpb=1877, bsz=2, num_updates=200, lr=2e-06, gnorm=2.725, train_wall=11, wall=222
2020-07-17 04:01:26 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-17 04:01:57 | INFO | train_inner | epoch 001:    300 / 2952000 loss=13.286, ppl=9988.85, wps=64.3, ups=0.03, wpb=2001, bsz=2, num_updates=300, lr=3e-06, gnorm=2.416, train_wall=11, wall=253
2020-07-17 04:02:28 | INFO | train_inner | epoch 001:    400 / 2952000 loss=11.931, ppl=3903.91, wps=61, ups=0.03, wpb=1900, bsz=2, num_updates=400, lr=4e-06, gnorm=5.17, train_wall=11, wall=284
2020-07-17 04:02:59 | INFO | train_inner | epoch 001:    500 / 2952000 loss=13.251, ppl=9747.4, wps=61, ups=0.03, wpb=1888, bsz=2, num_updates=500, lr=5e-06, gnorm=2.323, train_wall=11, wall=315
2020-07-17 04:03:30 | INFO | train_inner | epoch 001:    600 / 2952000 loss=12.786, ppl=7062.4, wps=61.1, ups=0.03, wpb=1891, bsz=2, num_updates=600, lr=6e-06, gnorm=2.155, train_wall=11, wall=346
2020-07-17 04:04:01 | INFO | train_inner | epoch 001:    700 / 2952000 loss=12.686, ppl=6590.07, wps=60.1, ups=0.03, wpb=1859, bsz=2, num_updates=700, lr=7e-06, gnorm=2.015, train_wall=11, wall=377
2020-07-17 04:04:32 | INFO | train_inner | epoch 001:    800 / 2952000 loss=12.712, ppl=6707.93, wps=63.9, ups=0.03, wpb=1988, bsz=2, num_updates=800, lr=8e-06, gnorm=3.707, train_wall=11, wall=408
2020-07-17 04:05:03 | INFO | train_inner | epoch 001:    900 / 2952000 loss=11.645, ppl=3203.19, wps=64.6, ups=0.03, wpb=1991, bsz=2, num_updates=900, lr=9e-06, gnorm=4.098, train_wall=11, wall=439
2020-07-17 04:05:34 | INFO | train_inner | epoch 001:   1000 / 2952000 loss=11.528, ppl=2953.08, wps=60.8, ups=0.03, wpb=1881, bsz=2, num_updates=1000, lr=1e-05, gnorm=1.523, train_wall=11, wall=470
2020-07-17 04:06:06 | INFO | train_inner | epoch 001:   1100 / 2952000 loss=10.505, ppl=1453.23, wps=62.5, ups=0.03, wpb=1971, bsz=2, num_updates=1100, lr=1.1e-05, gnorm=4.227, train_wall=12, wall=501
2020-07-17 04:06:36 | INFO | train_inner | epoch 001:   1200 / 2952000 loss=10.752, ppl=1725, wps=64.8, ups=0.03, wpb=1997, bsz=2, num_updates=1200, lr=1.2e-05, gnorm=2.649, train_wall=11, wall=532
2020-07-17 04:07:07 | INFO | train_inner | epoch 001:   1300 / 2952000 loss=11.373, ppl=2652.16, wps=62, ups=0.03, wpb=1912, bsz=2, num_updates=1300, lr=1.3e-05, gnorm=1.555, train_wall=11, wall=563
2020-07-17 04:07:38 | INFO | train_inner | epoch 001:   1400 / 2952000 loss=9.646, ppl=801.3, wps=62.4, ups=0.03, wpb=1932, bsz=2, num_updates=1400, lr=1.4e-05, gnorm=2.147, train_wall=11, wall=594
2020-07-17 04:08:09 | INFO | train_inner | epoch 001:   1500 / 2952000 loss=11.189, ppl=2334.9, wps=59.7, ups=0.03, wpb=1845, bsz=2, num_updates=1500, lr=1.5e-05, gnorm=1.252, train_wall=11, wall=625
2020-07-17 04:08:40 | INFO | train_inner | epoch 001:   1600 / 2952000 loss=11.251, ppl=2437.86, wps=56.3, ups=0.03, wpb=1743, bsz=2, num_updates=1600, lr=1.6e-05, gnorm=1.251, train_wall=11, wall=656
2020-07-17 04:09:12 | INFO | train_inner | epoch 001:   1700 / 2952000 loss=8.878, ppl=470.61, wps=64.1, ups=0.03, wpb=2011, bsz=2, num_updates=1700, lr=1.7e-05, gnorm=5.188, train_wall=12, wall=687
2020-07-17 04:09:43 | INFO | train_inner | epoch 001:   1800 / 2952000 loss=10.918, ppl=1935.07, wps=59.9, ups=0.03, wpb=1859, bsz=2, num_updates=1800, lr=1.8e-05, gnorm=5.364, train_wall=11, wall=718
2020-07-17 04:10:13 | INFO | train_inner | epoch 001:   1900 / 2952000 loss=10.954, ppl=1984.24, wps=65.4, ups=0.03, wpb=2017, bsz=2, num_updates=1900, lr=1.9e-05, gnorm=2.008, train_wall=11, wall=749
2020-07-17 04:10:44 | INFO | train_inner | epoch 001:   2000 / 2952000 loss=11.478, ppl=2851.54, wps=63.6, ups=0.03, wpb=1964, bsz=2, num_updates=2000, lr=2e-05, gnorm=2.558, train_wall=11, wall=780
2020-07-17 04:11:15 | INFO | train_inner | epoch 001:   2100 / 2952000 loss=9.209, ppl=591.91, wps=64.8, ups=0.03, wpb=2003, bsz=2, num_updates=2100, lr=2.1e-05, gnorm=4.174, train_wall=11, wall=811
2020-07-17 04:11:46 | INFO | train_inner | epoch 001:   2200 / 2952000 loss=10.191, ppl=1169.15, wps=64.8, ups=0.03, wpb=1999, bsz=2, num_updates=2200, lr=2.2e-05, gnorm=1.466, train_wall=11, wall=842
2020-07-17 04:12:17 | INFO | train_inner | epoch 001:   2300 / 2952000 loss=8.639, ppl=398.56, wps=62.9, ups=0.03, wpb=1942, bsz=2, num_updates=2300, lr=2.3e-05, gnorm=1.759, train_wall=11, wall=873

GPU log:
2020-07-17 04:08:14 | INFO | train_inner | epoch 001:    100 / 2952087 loss=16.348, ppl=83437.5, wps=8640.2, ups=4.66, wpb=1854, bsz=2, num_updates=100, lr=1e-06, gnorm=4.215, loss_scale=128, train_wall=22, wall=120
2020-07-17 04:08:36 | INFO | train_inner | epoch 001:    200 / 2952087 loss=14.736, ppl=27279.9, wps=8800.8, ups=4.69, wpb=1878.2, bsz=2, num_updates=200, lr=2e-06, gnorm=2.849, loss_scale=128, train_wall=21, wall=141
2020-07-17 04:08:57 | INFO | train_inner | epoch 001:    300 / 2952087 loss=13.761, ppl=13882.4, wps=8737.9, ups=4.68, wpb=1868.8, bsz=2, num_updates=300, lr=3e-06, gnorm=2.349, loss_scale=128, train_wall=21, wall=163
2020-07-17 04:09:19 | INFO | train_inner | epoch 001:    400 / 2952087 loss=13.199, ppl=9403.66, wps=8783.3, ups=4.66, wpb=1883.9, bsz=2, num_updates=400, lr=4e-06, gnorm=2.206, loss_scale=128, train_wall=21, wall=184
2020-07-17 04:09:40 | INFO | train_inner | epoch 001:    500 / 2952087 loss=12.839, ppl=7328.79, wps=8732.4, ups=4.66, wpb=1873.4, bsz=2, num_updates=500, lr=5e-06, gnorm=2.079, loss_scale=128, train_wall=21, wall=205
2020-07-17 04:10:02 | INFO | train_inner | epoch 001:    600 / 2952087 loss=12.426, ppl=5504.27, wps=8684.1, ups=4.66, wpb=1864.6, bsz=2, num_updates=600, lr=6e-06, gnorm=1.973, loss_scale=128, train_wall=21, wall=227
2020-07-17 04:10:23 | INFO | train_inner | epoch 001:    700 / 2952087 loss=11.955, ppl=3969.9, wps=8706.6, ups=4.66, wpb=1869.2, bsz=2, num_updates=700, lr=7e-06, gnorm=1.82, loss_scale=128, train_wall=21, wall=248
2020-07-17 04:10:44 | INFO | train_inner | epoch 001:    800 / 2952087 loss=11.481, ppl=2858.64, wps=8703.8, ups=4.65, wpb=1871.2, bsz=2, num_updates=800, lr=8e-06, gnorm=1.606, loss_scale=128, train_wall=21, wall=270
2020-07-17 04:11:00 | INFO | fairseq.trainer | NOTE: overflow detected, setting loss scale to: 64.0
2020-07-17 04:11:06 | INFO | train_inner | epoch 001:    901 / 2952087 loss=11.042, ppl=2108.45, wps=8672.3, ups=4.6, wpb=1884.5, bsz=2, num_updates=900, lr=9e-06, gnorm=1.521, loss_scale=110, train_wall=22, wall=292
2020-07-17 04:11:28 | INFO | train_inner | epoch 001:   1001 / 2952087 loss=10.691, ppl=1653.52, wps=8731.1, ups=4.64, wpb=1881.1, bsz=2, num_updates=1000, lr=1e-05, gnorm=1.324, loss_scale=64, train_wall=21, wall=313
2020-07-17 04:11:49 | INFO | train_inner | epoch 001:   1101 / 2952087 loss=10.477, ppl=1425.28, wps=8661.6, ups=4.64, wpb=1865.9, bsz=2, num_updates=1100, lr=1.1e-05, gnorm=1.29, loss_scale=64, train_wall=21, wall=335
2020-07-17 04:12:11 | INFO | train_inner | epoch 001:   1201 / 2952087 loss=10.289, ppl=1251.5, wps=8674.2, ups=4.64, wpb=1869, bsz=2, num_updates=1200, lr=1.2e-05, gnorm=1.272, loss_scale=64, train_wall=21, wall=356
2020-07-17 04:12:32 | INFO | train_inner | epoch 001:   1301 / 2952087 loss=10.146, ppl=1132.95, wps=8659.2, ups=4.64, wpb=1865.1, bsz=2, num_updates=1300, lr=1.3e-05, gnorm=1.263, loss_scale=64, train_wall=21, wall=378
2020-07-17 04:12:54 | INFO | train_inner | epoch 001:   1401 / 2952087 loss=10.017, ppl=1035.98, wps=8641.4, ups=4.64, wpb=1862.1, bsz=2, num_updates=1400, lr=1.4e-05, gnorm=1.248, loss_scale=64, train_wall=21, wall=399
2020-07-17 04:13:16 | INFO | train_inner | epoch 001:   1501 / 2952087 loss=9.973, ppl=1004.81, wps=8698.6, ups=4.64, wpb=1875.2, bsz=2, num_updates=1500, lr=1.5e-05, gnorm=1.287, loss_scale=64, train_wall=21, wall=421
2020-07-17 04:13:37 | INFO | train_inner | epoch 001:   1601 / 2952087 loss=9.916, ppl=965.88, wps=8645.5, ups=4.64, wpb=1863.9, bsz=2, num_updates=1600, lr=1.6e-05, gnorm=1.31, loss_scale=64, train_wall=21, wall=442
2020-07-17 04:13:59 | INFO | train_inner | epoch 001:   1701 / 2952087 loss=9.782, ppl=880.15, wps=8619.2, ups=4.63, wpb=1860.5, bsz=2, num_updates=1700, lr=1.7e-05, gnorm=1.31, loss_scale=64, train_wall=21, wall=464
2020-07-17 04:14:20 | INFO | train_inner | epoch 001:   1801 / 2952087 loss=9.826, ppl=907.57, wps=8653.1, ups=4.64, wpb=1863.2, bsz=2, num_updates=1800, lr=1.8e-05, gnorm=1.391, loss_scale=64, train_wall=21, wall=486
2020-07-17 04:14:42 | INFO | train_inner | epoch 001:   1901 / 2952087 loss=9.69, ppl=826.16, wps=8686.3, ups=4.64, wpb=1871.7, bsz=2, num_updates=1900, lr=1.9e-05, gnorm=1.407, loss_scale=64, train_wall=21, wall=507
2020-07-17 04:15:03 | INFO | train_inner | epoch 001:   2001 / 2952087 loss=9.64, ppl=797.68, wps=8702.3, ups=4.64, wpb=1875.5, bsz=2, num_updates=2000, lr=2e-05, gnorm=1.389, loss_scale=64, train_wall=21, wall=529
2020-07-17 04:15:25 | INFO | train_inner | epoch 001:   2101 / 2952087 loss=9.503, ppl=725.43, wps=8631.3, ups=4.64, wpb=1860.1, bsz=2, num_updates=2100, lr=2.1e-05, gnorm=1.434, loss_scale=64, train_wall=21, wall=550
2020-07-17 04:15:46 | INFO | train_inner | epoch 001:   2201 / 2952087 loss=9.366, ppl=660.07, wps=8743.7, ups=4.64, wpb=1883.9, bsz=2, num_updates=2200, lr=2.2e-05, gnorm=1.426, loss_scale=64, train_wall=21, wall=572
2020-07-17 04:16:08 | INFO | train_inner | epoch 001:   2301 / 2952087 loss=9.4, ppl=675.6, wps=8639.4, ups=4.64, wpb=1861.4, bsz=2, num_updates=2300, lr=2.3e-05, gnorm=1.379, loss_scale=64, train_wall=21, wall=593

Have you seen anything similar? @myleott 
Any thoughts?
Thanks

**Comment by myleott** [comment]
What training command did you run? Are you using bfloat16 by any chance? If so, I highly recommend trying without it.

**Comment by kkissmart** [comment]
i tried both with and without

here is the command:

python train.py /home/xwu/data-bin --encoder-normalize-before --decoder-normalize-before --arch mbart_base --layernorm-embedding --task multilingual_denoising --criterion cross_entropy --dataset-impl mmap  --optimizer adam --adam-eps 1e-06 --adam-betas '(0.9, 0.98)' --lr-scheduler polynomial_decay --lr 1e-04 --min-lr -1 --warmup-updates 5000 --total-num-update 500000 --dropout 0.0 --attention-dropout 0.0 --weight-decay 0.0 --max-tokens 4104  --seed 222 --log-format simple --log-interval 100  --add-lang-token --no-whole-word-mask-langs JA  --mask 0.35 --permute-sentences 1.0 --mask-length span-poisson --replace-length 1 --rotate 0.0 --max-source-positions 1026 --max-target-positions 1026 --tokens-per-sample 1026 --sample-break-mode complete  --save-interval-updates 5000 --skip-invalid-size-inputs-valid-test --langs EN,JA  --multilang-sampling-alpha 0.7 --distributed-world-size 8  --tpu  --no-bos  --save-dir $HOME  --num-buckets 2 --no-input-eos --share-all-embeddings  --max-update 500000

**Comment by myleott** [comment]
Hmm, I've never tried mbart with TPUs. One observation: it doesn't seem like those two logs you shared are really comparable. The bsz on the GPU log is bsz=2, but on the TPU log is bsz=128. The learning rates also seem to be different.

**Comment by kkissmart** [comment]
@myleott @taylanbil 
I updated the logs to make sure they are apple2apple comparison. 

Thanks!
